src/functions.php

```diff
@@ -342,33 +342,67 @@ function _checkTypehint(callable $callback, \Throwable $reason): bool
         return true;
     }
 
-    $type = $parameters[0]->getType();
+    $expectedException = $parameters[0];
 
-    if (!$type) {
-        return true;
-    }
+    // PHP before v8 used an easy API:
+    if (\PHP_VERSION_ID < 70100 || \defined('HHVM_VERSION')) {
+        if (!$expectedException->getClass()) {
+            return true;
+        }
 
-    $types = [$type];
+        return $expectedException->getClass()->isInstance($reason);
+    }
```

Comment:
Promise v3 is PHP 7.1+ without HHVM, so it looks like this can be simplified :+1: